src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -929,15 +941,15 @@ private void VisitPattern(BoundExpression expression, TypeSymbolWithAnnotations
 
             // https://github.com/dotnet/roslyn/issues/29873 We should only report such
             // diagnostics for locals that are set or checked explicitly within this method.
-            if (!expressionResultType.IsNull && expressionResultType.IsNullable == false && isNull == true)
+            if (!expressionResultType.IsNull && expressionResultType.ValueCanBeNull() == false && whenTrue == NullableAnnotation.NullableBasedOnAnalysis)
             {
                 ReportDiagnostic(ErrorCode.HDN_NullCheckIsProbablyAlwaysFalse, pattern.Syntax);
             }
 
-            if (slot > 0 && isNull != null)
+            if (slot > 0 && whenTrue != NullableAnnotation.Unknown)
```

Comment:
>whenTrue != NullableAnnotation.Unknown [](start = 28, length = 38)

Consider asserting this instead.

Reply:
> Consider asserting this instead.

Coming in the next commit.

---
In reply to: [231333178](https://github.com/dotnet/roslyn/pull/30913#discussion_r231333178) [](ancestors = 231333178)